### PR TITLE
Add Labels And Annotations From Pods To ServiceInstance Metadata

### PR DESCRIPTION
**Pull request by weihubeats**
https://github.com/spring-cloud/spring-cloud-kubernetes/issues/1162

### CODE REVIEW

**Review by ryanjbaxter — CHANGES_REQUESTED** [review]
I prefer if we take the code which collects the service metadata and the pod metadata and separate that out into its own method.  This is getting a bit complex with all the code in one method.  Also you want to respect whether the add labels and annotations preference is set.

**Comment by ryanjbaxter** [comment]
Also please fix the checkstyle errors

**Comment by weihubeats** [comment]
ok

File: spring-cloud-kubernetes-client-discovery/src/main/java/org/springframework/cloud/kubernetes/client/discovery/KubernetesInformerDiscoveryClient.java [hunk 1]
```diff
@@ -145,39 +159,61 @@ private Stream<ServiceInstance> getServiceInstanceDetails(V1Service service, Str
...
+	private Map<String, String> initPodMetaData(V1EndpointAddress addr, Map<String, String> metadata, Map<String, V1ObjectMeta> podMap) {
+		Map<String, String> newMetaData = new HashMap<>(metadata);
+		if (!CollectionUtils.isEmpty(podMap)) {
+			V1ObjectReference targetRef = addr.getTargetRef();
+			if (!ObjectUtils.isEmpty(targetRef)) {
+				String name = targetRef.getName();
+				if (!ObjectUtils.isEmpty(name)) {
+					Optional.ofNullable(podMap.get(name)).ifPresent(podMetaData -> {
+						if (!CollectionUtils.isEmpty(podMetaData.getLabels())) {
+							newMetaData.putAll(podMetaData.getLabels());
```
Comment (ryanjbaxter) [inline spring-cloud-kubernetes-client-discovery/src/main/java/org/springframework/cloud/kubernetes/client/discovery/KubernetesInformerDiscoveryClient.java:210]: is this just adding labels, what about annotations?

File: spring-cloud-kubernetes-client-discovery/src/main/java/org/springframework/cloud/kubernetes/client/discovery/KubernetesInformerDiscoveryClient.java [hunk 2]
```diff
@@ -145,39 +159,61 @@ private Stream<ServiceInstance> getServiceInstanceDetails(V1Service service, Str
...
+				if (this.properties.includeNotReadyAddresses()
+					&& !CollectionUtils.isEmpty(subset.getNotReadyAddresses())) {
+					addresses.addAll(subset.getNotReadyAddresses());
+				}
+
+				final int port = findEndpointPort(endpointPorts, primaryPortName, serviceId);
+				return addresses.stream()
+					.map(addr -> new DefaultKubernetesServiceInstance(
+						addr.getTargetRef() != null ? addr.getTargetRef().getUid() : "", serviceId,
+						addr.getIp(), port, properties.initPodMetaData() ? initPodMetaData(addr, metadata, finalPodMap) : metadata, secured, service.getMetadata()
```
Comment (ryanjbaxter) [inline spring-cloud-kubernetes-client-discovery/src/main/java/org/springframework/cloud/kubernetes/client/discovery/KubernetesInformerDiscoveryClient.java:192]: IDK that we need a separate property, we already have a property to enable adding annotations and labels

File: spring-cloud-kubernetes-client-discovery/src/main/java/org/springframework/cloud/kubernetes/client/discovery/KubernetesInformerDiscoveryClient.java (hunk 1 above)
Comment (weihubeats) [reply spring-cloud-kubernetes-client-discovery/src/main/java/org/springframework/cloud/kubernetes/client/discovery/KubernetesInformerDiscoveryClient.java:212]: The metadata obtained here is the metadata of the pod, which should belong to metadata by definition

File: spring-cloud-kubernetes-client-discovery/src/main/java/org/springframework/cloud/kubernetes/client/discovery/KubernetesInformerDiscoveryClient.java (hunk 2 above)
Comment (weihubeats) [reply spring-cloud-kubernetes-client-discovery/src/main/java/org/springframework/cloud/kubernetes/client/discovery/KubernetesInformerDiscoveryClient.java:194]: The metadata obtained here is the metadata of the pod, which should belong to metadata by definition

File: spring-cloud-kubernetes-client-discovery/src/main/java/org/springframework/cloud/kubernetes/client/discovery/KubernetesInformerDiscoveryClient.java (hunk 1 above)
Comment (weihubeats) [reply spring-cloud-kubernetes-client-discovery/src/main/java/org/springframework/cloud/kubernetes/client/discovery/KubernetesInformerDiscoveryClient.java:212]: done

File: spring-cloud-kubernetes-client-discovery/src/main/java/org/springframework/cloud/kubernetes/client/discovery/reactive/KubernetesInformerReactiveDiscoveryClient.java [hunk 3]
```diff
@@ -36,15 +37,15 @@
...
-	private KubernetesInformerDiscoveryClient kubernetesDiscoveryClient;
+	private final KubernetesInformerDiscoveryClient kubernetesDiscoveryClient;
 
 	public KubernetesInformerReactiveDiscoveryClient(KubernetesNamespaceProvider kubernetesNamespaceProvider,
-			SharedInformerFactory sharedInformerFactory, Lister<V1Service> serviceLister,
-			Lister<V1Endpoints> endpointsLister, SharedInformer<V1Service> serviceInformer,
-			SharedInformer<V1Endpoints> endpointsInformer, KubernetesDiscoveryProperties properties) {
+		SharedInformerFactory sharedInformerFactory, Lister<V1Service> serviceLister,
+		Lister<V1Endpoints> endpointsLister, SharedInformer<V1Service> serviceInformer, Lister<V1Pod> podLister,
+		SharedInformer<V1Endpoints> endpointsInformer, KubernetesDiscoveryProperties properties) {
```
Comment (ryanjbaxter) [inline spring-cloud-kubernetes-client-discovery/src/main/java/org/springframework/cloud/kubernetes/client/discovery/reactive/KubernetesInformerReactiveDiscoveryClient.java:45]: Same here we need a new constructor

**Review by ryanjbaxter — COMMENTED** [review]
A couple of other things:

1.) We would need to add the same functionality to the Fabric8 implementation
2.) I am still unsure of what will happen when labels/annotations from the pod and service conflict.  Seems like they will override either other, last one added will win.  I am not sure how we could resolve this.

**Comment by weihubeats** [comment]
> A couple of other things:
> 
> 1.) We would need to add the same functionality to the Fabric8 implementation 2.) I am still unsure of what will happen when labels/annotations from the pod and service conflict. Seems like they will override either other, last one added will win. I am not sure how we could resolve this.

I think there is no good way to solve the conflict, but only through the user to control themselves, because the metadata in the ServiceInstance is just a Map<String, String>

**Comment by ryanjbaxter** [comment]
We still need documentation

**Comment by wind57** [comment]
correlating the the address to the pod is not complicated, for example:

```
kubectl get -o json endpoints spring-cloud-kubernetes-client-configmap-event-reload
{
    "apiVersion": "v1",
    "kind": "Endpoints",
    "metadata": {
        "annotations": {
            "endpoints.kubernetes.io/last-change-trigger-time": "2023-02-21T14:01:05Z"
        },
        "creationTimestamp": "2023-02-21T14:00:55Z",
        "labels": {
            "app": "spring-cloud-kubernetes-client-configmap-event-reload"
        },
        "name": "spring-cloud-kubernetes-client-configmap-event-reload",
        "namespace": "default",
        "resourceVersion": "613",
        "uid": "2d017d3a-a18a-4ddb-ae2f-05c3f61ab117"
    },
    "subsets": [
        {
            "addresses": [
                {
                    "ip": "10.42.0.7",
                    "nodeName": "c577996ade89",
                    "targetRef": {
                        "kind": "Pod",
                        "name": "spring-cloud-kubernetes-client-configmap-deployment-event-jw5jz",
                        "namespace": "default",
                        "uid": "a499484e-0d8c-4668-be9c-f4df52c550d3"
                    }
                }
            ],
            "ports": [
                {
                    "name": "http",
                    "port": 8080,
                    "protocol": "TCP"
                }
            ]
        }
    ]
}
/ #
```

notice that each address (if it has a `targetRef`), has : 


```
                    "targetRef": {
                        "kind": "Pod",
                        "name": "spring-cloud-kubernetes-client-configmap-deployment-event-jw5jz",
                        "namespace": "default",
                        "uid": "a499484e-0d8c-4668-be9c-f4df52c550d3"
                    }
```

As to returning the metadata from pods, we could introduce a new record in `DefaultKubernetesServiceInstance`, for example `PodMetadata` that would contain things like pod-name and pod-labels. Of course you have to opt it, to get such data.

@ryanjbaxter if you agree, I would continue this at some point in time.

**Comment by wind57** [comment]
I've added such functionality to the fabric8 client, pending review [here](https://github.com/spring-cloud/spring-cloud-kubernetes/pull/1254). I will, of course, add such an option to the k8s-client too later.

**Comment by ryanjbaxter** [comment]
Addressed in https://github.com/spring-cloud/spring-cloud-kubernetes/pull/1254

**Pull request closed without merging**